server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4302,22 +4414,29 @@ public void afterRefresh(boolean didRefresh) {
     /**
      * Reset the current engine to a new one.
      *
-     * Calls {@link Engine#prepareForEngineReset()} on the current engine, then closes it, and loads a new engine without
+     * Calls {@link Engine#beforeReset()} on the current engine, then closes it, and loads a new engine without
      * doing any translog recovery.
      *
      * In general, resetting the engine should be done with care, to consider any in-progress operations and listeners.
      * At the moment, this is implemented in serverless for a special case that ensures the engine is prepared for reset.
      */
-    public void resetEngine() {
+    private void resetEngine(boolean mutability) {
         assert Thread.holdsLock(mutex) == false : "resetting engine under mutex";
         assert waitForEngineOrClosedShardListeners.isDone();
         try {
-            synchronized (engineMutex) {
+            engineLock.writeLock().lock(); // might already be held
+            try {
                 verifyNotClosed();
-                getEngine().prepareForEngineReset();
-                var newEngine = createEngine(newEngineConfig(replicationTracker));
-                IOUtils.close(currentEngineReference.getAndSet(newEngine));
-                onNewEngine(newEngine);
+                if (currentEngine.isMutable() != mutability) {
```

Comment:
Similar to previous comment (and Iraklis') can we make this more like `currentEngine.pendingReset()`?